### DISCUSSION TITLE
How to run make REST calls to Azure Resource Manager (ARM) inside Github Workflow?

### DISCUSSION THREAD

**Discussion by siegfried01** [community/community#67346] [Other Feature Feedback, Questions, & Ideas] [answered]
### Background

I'm cross posting from [https://learn.microsoft.com/en-us/answers/questions/1355574/how-to-run-make-rest-calls-to-azure-resource-manag?source=docs](url) partially because I have not received any responses in a while (maybe that post has too much history to wade thru?) and partially because this forum might be more appropriate.

### Powershell Script works great from personal laptop

I have a working powershell script inspired by [https://techcommunity.microsoft.com/t5/azure-paas-blog/import-azure-function-app-to-azure-api-management/ba-p/2594810](url) and I would like to run it in a Github workflow.  This script uses a REST call to fetch the azure function app keys and I want to add another REST call to fetch the metadata from the function app. Presently I cannot make the first REST call work inside the github work flow. Everything (include both REST calls) works fine when running from my personal laptop.

Here is a copy of my attempt: [https://github.com/siegfried01/CrewTaskManagerTempPublic/blob/main/jac3sukjdrxzi-func-CrewTaskMgrAuthSvcs-Sat-Sep-16-2023-14.yml#L66](url)

(The original here but that might change if I do another push: [https://github.com/siegfried01/HelloAzureADAuthenticatedFunc/blob/master/.github/workflows/jac3sukjdrxzi-func-CrewTaskMgrAuthSvcs.yml#L66](url))

### Error Messages from Github Workflow: "Interactive authentication is needed"

I'm getting these errors (see [https://github.com/siegfried01/HelloAzureADAuthenticatedFunc/actions/runs/6209491656/job/16856766981](url)) from the "az rest" with a bearer token.

    ERROR: Unauthorized(***"error":***"code":"InvalidAuthenticationToken","message":"The access token is invalid."***)
    Interactive authentication is needed. Please run:
    az logout
    az login

Well, as you can see, I have already logged in twice (as suggested by ChatGPT) using the actions/login@v1 and again using "az login".

### Additional Background Details:

I have created a service principal using these commands

    export id=`az group show --name $rg --query 'id' --output tsv`
    az ad sp create-for-rbac --name $sp --sdk-auth --role contributor --scopes $id

I have created github secrets called AZURE_CREDENTIALS (using the output from the above command), AZURE_CLIENT_ID, AZURE_CLIENT_SECRET & AZURE_TENANT_ID.

I have (redundantly?) granted the service principal contributor access to the function app:

    az role assignment create --assignee $clientId --role Contributor --scope "/subscriptions/$subscriptionId/resourceGroups/$rg/providers/Microsoft.Web/sites/$functionApp"

Please guide me in making an authorized REST call from inside a Github workflow.

Thanks

Siegfried

**Comment by Kaushikfortis** [comment, marked as answer]
Create a GitHub Workflow:

In your repository, create or edit an existing GitHub Workflow file (e.g., .github/workflows/azure-rest-api.yml). Define your workflow steps to make REST calls to the Azure Resource Manager API.

Here's an example of a simple GitHub Workflow that makes an Azure REST API call using cURL:

name: Azure REST API Workflow

on:
  push:
    branches:
      - main

jobs:
  deploy:
    runs-on: ubuntu-latest

    steps:
      - name: Checkout Repository
        uses: actions/checkout@v2

      - name: Set up Azure CLI
        uses: azure/login@v1
        with:
          creds: ${{ secrets.AZURE_CLIENT_SECRET }}
          enable-AzPSSession: true

      - name: Make Azure REST API Call
        run: |
          # Replace with your Azure REST API call using cURL or other HTTP client.
          curl -X GET -H "Authorization: Bearer $AZURE_HTTP_API_AUTH_TOKEN" \
            https://management.azure.com/subscriptions/{subscription_id}/resourceGroups?api-version=2021-04-01


And then Customize API calls and commit& Push.

**Reply by siegfried01** [reply]
With some modifications, this works! I wrongly assumed that AZURE_HTTP_API_AUTH_TOKEN was built-in. It is not!

I had to add: $AZURE_HTTP_API_AUTH_TOKEN=(az account get-access-token --query 'accessToken' -o tsv)

Now it works! Thank you!
